### DISCUSSION TITLE
I've been playing with Dark a little bit, it seems pretty powerful although there's certainly a learning curve

### DISCUSSION THREAD

**Discussion by pbiggar** [darklang/dark#4308] [Feedback]
A user says:
>I've been playing with Dark a little bit, it seems pretty powerful although there's certainly a learning curve so I've found my progress to be relatively slow.
I find myself blocked trying to do things that I would have expected to be pretty basic. Here's the last one, for example:
>
>I have a data store with 2 entries, called Users. I added an HTTP endpoint with /users/:name to fetch the data for a user, by name.
>Writing the endpoint like this worked fine:
>DB::get name Users
>
>I can see that there's a little v2 that appears right after DB::get, and from the docs I could find the definition as:
>DB::get_v2(Str key, Datastore table) -> Option
>
>I don't understand how my implementation works if the value was an optional. So I figured if it's an Option type, I could add support for error handling, returning either the User data, or an error message.
>So I wrote:
>
>let opt = DB::get name Users
>Option::withDefault opt {
>                          error : "no such user"
>                        }
>which produced:
>Application error: the executed program was invalid. This problem can be resolved by the application's author by resolving the invalid code (often a type error).
>
>I also tried it as:
>let opt = DB::get name Users
>opt.withDefault { error: "no such user" } 
>But the editor wouldn't even let me type that.
>
>I can see previous executions of the endpoint, which is really cool, and it really seems like the value I was getting from DB::get(v2) was a User object with the fields I had defined in the data store (e.g. name, email…)
That means I'm not getting an Option back like the docs say, and that's why I was initially puzzled that just returning the value I got from DB::get would even work.
>
>So… are the docs just wrong or am I missing something in my approach to such a simple problem?
I'll keep trying for a little bit and will hop on Slack to ask if I can't figure it out, or if it's not obvious to you either what it is that I'm doing wrong.

**Comment by StachuDotNet** [comment]
I'm guessing that this is old feedback, backfilled to this Discussion, but regardless:

Note that `Option.withDefault` does not _modify_ the argument's `option`, but rather takes an `Option<T>` and returns a `T` - if the `option` was a Some/Just, it'd return that value, otherwise if it's a None it'd return the default value passed in. It seems both of your attempts are trying to modify the value in-place.

So, something like this should yield better results:

`let opt = DB::get name Users |> Option.withDefault { ... default value ... }`

or if you want to break it into 2 lines, without a pipe:

```
let userMaybe = DB::get name Users
let user = Option.withDefault userMaybe { ... default value ... }
```

**Reply by pbiggar** [reply]
I read this as a user who is used to object-oriented program trying to use a dot where a dot is not valid in our editor. Perhaps pressing "dot" on an "object" should have a better path to success.

**Reply by pbiggar** [reply]
I think the ErrorRail is also contributing to the confusion.
